### PR TITLE
Modify device check in capturable optimizer to support more devices

### LINKED ISSUE #124830: Add privateuse1 check on capturable optimizer
### 🚀 The feature, motivation and pitch

We develop a tool similar to CUDAGraph on our own custom device (using dispatch key PrivateUse1), but capturable optimizer hardcode cuda like below:
```
# If compiling, the compiler will handle cudagraph checks, see note [torch.compile x capturable]
        if not torch._utils.is_compiling() and capturable:
            assert (param.is_cuda and step_t.is_cuda) or (
                param.is_xla and step_t.is_xla
            ), "If capturable=True, params and state_steps must be CUDA or XLA tensors."
```

Can I add privateuse1 check in these checks, or any other more general solutions?

### Alternatives

_No response_

### Additional context

_No response_
```[tasklist]
### Tasks
```

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang

### PR DESCRIPTION
**Pull request by accelerate321**
Fixes #124830 

Modify device check in capturable optimizer to support more device

### CODE REVIEW

File: torch/optim/adadelta.py [hunk 1]
```diff
@@ -310,8 +310,8 @@ def _single_tensor_adadelta(
     # If compiling, the compiler will handle cudagraph checks, see note [torch.compile x capturable]
     if not torch._utils.is_compiling() and capturable:
         assert all(
-            p.is_cuda and step.is_cuda for p, step in zip(params, state_steps)
-        ), "If capturable=True, params and state_steps must be CUDA tensors."
+            p.device == step.device for p, step in zip(params, state_steps)
```
Comment (janeyx99) [inline torch/optim/adadelta.py:313]: Now that I'm seeing that this is less safe than before, I think we should add a p's device type is in a _supports_capturable_devices list = ["cuda", "xla", "privateuse1"]. The list can live in optimizer.py for now.

Comment (accelerate321) [reply torch/optim/adadelta.py:313]: @janeyx99 
1. I find that some optimizers do not check xla now，is it ok if I add xla in  _supports_capturable_devices?
2. I will change to the below code, is it fine with you?
```
if not torch._utils.is_compiling() and capturable:
        assert all(
            p.device.type == step.devcie.type
            and p.device.type in _supports_capturable_devices_list
            for p, step in zip(params, state_steps)
        ), f"If capturable=True, params and state_steps must be on supported devices: {_supports_capturable_devices_list}."
```

**Review by janeyx99 — COMMENTED** [review]
Ah, it looks like none of our foreach (`multi_tensor_*`) impls is supported by XLA. To maintain correctness with prior code, could you add a way in `_get_capturable_supported_devices` to take in `supports_xla: bool` and don't return the list with XLA when it's set to False?

**Comment by accelerate321** [comment]
@janeyx99 I add `supports_xla: bool` argument and keep the xla check same with prior code, you can see whether it is right

**Review by janeyx99 — APPROVED** [review]
looks great let's ship it!

**Comment by janeyx99** [comment]
Please fix lint!

**Comment by accelerate321** [comment]
@janeyx99 I have fixed the lint, but still failed in CI due to the following reason in test_functional_optim.py
`Python builtin <built-in function _get_privateuse1_backend_name> is currently not supported in Torchscript:`
Do you have some ideas to fix it?

**Comment by accelerate321** [comment]
@janeyx99 I do not care about jit.script now,  can I change the code to the following to bypass the failure in CI, or any other solution？
```
def _get_capturable_supported_devices(supports_xla: bool = True) -> List[str]:
    r"""Return the device type list that supports capturable optimizer."""
    capturable_supported_devices = ["cuda"]
    if not torch.jit.is_scripting():
        capturable_supported_devices.append(torch._C._get_privateuse1_backend_name())
    if supports_xla:
        capturable_supported_devices.append("xla")
    return capturable_supported_devices
```

**Comment by janeyx99** [comment]
@accelerate321 that is totally fine!

**Comment by accelerate321** [comment]
@janeyx99 can you merge it? I have no authority.

**Pull request closed without merging**